src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlParameter.cs

```diff
@@ -331,9 +330,13 @@ string xmlSchemaCollectionName
             SourceVersion = sourceVersion;
             SourceColumnNullMapping = sourceColumnNullMapping;
             Value = value;
-            XmlSchemaCollectionDatabase = xmlSchemaCollectionDatabase;
-            XmlSchemaCollectionOwningSchema = xmlSchemaCollectionOwningSchema;
-            XmlSchemaCollectionName = xmlSchemaCollectionName;
+            if (!string.IsNullOrEmpty(xmlSchemaCollectionDatabase) || !string.IsNullOrEmpty(xmlSchemaCollectionOwningSchema) || !string.IsNullOrEmpty(xmlSchemaCollectionName))
+            {
+                EnsureXmlSchemaCollectionExists();
```

Comment:
If you don't have any specific reason for using property instead of a field and calling this function, add a property for `_xmlSchemaCollection`, please.

Reply:
There is a reason. If you have a property you would have it auto create the `_xmlSchemaCollection` field if it was null and then pass it back. That is harmless. The problem is that people then get used to using property accessor for `XmlSchemaCollection` and you will end up with code like `if (!string.IsNullOrEmpty(XmlSchemaCollection.Database)` which looks harmless but results in the schema container being created just to check if it's empty which isn't good. 

So I chose this pattern even though it's a little more work so that access is always through the field and anyone modifying the code can see that they need to call the create function before they assume that it's non-null.

Reply:
I could add a property called `EnsureXmlSchemaCollection` and then we end up with the property access pattern used like:
```csharp
public string XmlSchemaCollectionDatabase
        {
            get => _xmlSchemaCollection?.Database ?? string.Empty;
            set
            {
                EnsureXmlSchemaCollection.Database = value;
            }
        }
```

what do you think?

Reply:
With this approach, I think except in ctor you can use the property `EnsureXmlSchemaCollection` instead of `_xmlSchemaCollection` to assign a value or call a function.

Reply:
Yup. Done. It looks much neater in general. I've gone with a function over a property because of the api guidance that trivial operations should be properties but anything which has non-obvious side effects should really be a function. See what you think.